### ISSUE TITLE
Some questions about the labels and image encoders

### ISSUE THREAD

**Issue by LeeXiaosong** [FahimZaman/LDSeg#4]
Dear author, 

you have done a very nice work in the field of image segmentation. Thank you for your contribution to the open source community.
I have a question about the encoders for labels and images. Do you directly use existing frameworks and pre trained models for these two encoders? And freeze the encoder and decoder parts in training the diffusion model? If not, how is this part trained? 
I tried to apply your framework in other tasks, but my input is not an image. It requires training a shallow feature extraction network to obtain a multi-channel pseudo image.

Looking forward to your reply.

**Comment by FahimZaman** [comment]
The model is trained end to end without any pretrained weights. For segmentation purpose, there are two losses, one is the denoiser loss in the latent space, the other one is the segmentation loss in the pixel space. A forward pass in training phase starts from the label and image encoders, then the noisy latent of label is obtained using the Gaussian block which goes through denoiser, finally the cleaned latent of label goes through the decoder to produce the segmentation. In your case, I think you can add the feature extractor before the image encoder; or you can replace the image encoder with the feature extractor and concatenate the features with the label latent. But you have to be careful with pseudo image size (maybe reshaping with interpolation?).

**Comment by LeeXiaosong** [comment]
> The model is trained end to end without any pretrained weights. For segmentation purpose, there are two losses, one is the denoiser loss in the latent space, the other one is the segmentation loss in the pixel space. A forward pass in training phase starts from the label and image encoders, then the noisy latent of label is obtained using the Gaussian block which goes through denoiser, finally the cleaned latent of label goes through the decoder to produce the segmentation. In your case, I think you can add the feature extractor before the image encoder; or you can replace the image encoder with the feature extractor and concatenate the features with the label latent. But you have to be careful with pseudo image size (maybe reshaping with interpolation?).

Dear Zaman:

Thank for your reply!
based on your explanation, can I think of LDSeg as the result of a single, end-to-end network training? But, in other diffusion models, the training is usually split into two stages. The first stage is to train something like an encoder-decoder network, this is used for feature extraction and restoration; and the second stage is to train the denoising network. In LDSeg, if we train the label-encoder, image-encoder, label-decoder, and denoiser all at one process, can it be sure that the encoder learns an effective feature extraction process?

**Comment by FahimZaman** [comment]
Yes, you are right, this is the main difference of LDSeg and other conditional diffusion models as it is a true end-to-end trained model. From my experience and experiments, I found out that both the encoders can accurately learn meaningful features even with the end-to-end pipeline, at least for the segmentation application, otherwise I would not be able to get state-of-the-art segmentation accuracy in different dataset. Also, the residual noise in the label latents kind of forces the label encoder and decoder to act similar to a VAE. As a result, you do not need 2 step training process, but if you do want to train in two steps, a simple autoencoder for the label image may not be sufficient, you may need to train a VAE or VQVAE to get better result.

**Comment by LeeXiaosong** [comment]
> Yes, you are right, this is the main difference of LDSeg and other conditional diffusion models as it is a true end-to-end trained model. From my experience and experiments, I found out that both the encoders can accurately learn meaningful features even with the end-to-end pipeline, at least for the segmentation application, otherwise I would not be able to get state-of-the-art segmentation accuracy in different dataset. Also, the residual noise in the label latents kind of forces the label encoder and decoder to act similar to a VAE. As a result, you do not need 2 step training process, but if you do want to train in two steps, a simple autoencoder for the label image may not be sufficient, you may need to train a VAE or VQVAE to get better result.

Thank you Zaman! 
Training both enc-dec, and diffusion networks at the same time sounds very cool. In LDSeg, it seems that there is no detailed training set mentioned. I would like to ask how many epochs need to set or how much time you spend for training this pipeline to obtain results similar to those in your paper? Besides, LDSeg seems to be based on the TensorFlow architecture. If I want to transfer it to PyTorch, do you have any recommended routines or example of diffusion framework?
I'm very happy for discussing diffusion model with you😄

**Comment by FahimZaman** [comment]
If you look at the cfg.ini file, it contains all the default training/inference parameters. But, number of epoch can vary depending on your dataset, as for some data the model converges pretty quickly. The encoders and decoder are standard CNN with/without attention mechanism, which I think can be easily adapted in PyTorch framework. For the denoiser, you can look at the original DDPM paper implementation in PyTorch (Ho et al.). Overall, I think it should be easily adaptable in PyTorch framework.
